### PR TITLE
Use a .dat folder to resume links and RC2 API (#476, #484)

### PR DESCRIPTION
**Pull request by joehand**
### General Changes
* Uses `.dat` folder to resume links in the given directory (share & download)
  * Resuming a share only adds *new files* not in archive
  * **Does not support** updated files
* Keep download connections open for both live & static (no option to close).
* Updates tests
* Use leveldb

### API Changes
* Updates API to use proposed RC2 version (see usage.txt)
* Removes most options
* Add `--list` option
* Rename `--static` to `--snapshot` 
* Make live default

### Output Changes
* Update wording around dat link:
screen shot 2016-06-16 at 10 02 30
* Tell user when a Dat is resumed
* Show download/share folder

### CODE REVIEW

**Comment by mafintosh** [comment]
@joehand "Does not support updated files" - I need to fix this right?

**Comment by mafintosh** [comment]
Why is there no flag to exit on content downloaded? That sounds very useful when scripting stuff.

**Comment by mafintosh** [comment]
I actually like a separate cmd that fetches and exits more than a flag now that i think about it.

**Comment by joehand** [comment]
> "Does not support updated files" - I need to fix this right?

@mafintosh ya, we need to store the mtimes in hyperdrive for this. Once we have that, will be straightforward for me to check.
 
> Why is there no flag to exit on content downloaded? That sounds very useful when scripting stuff.

Oh ya, didn't think about that use. I can add it.

**Comment by joehand** [comment]
> I actually like a separate cmd that fetches and exits more than a flag now that i think about it.

Ya I think that'd make sense - but do we want to support both of those? `dat checkout <link>`? 

We could change `dat get <link>` to something more specific too. `dat join <link>`? Not sure what describes it properly (downloading, connecting, and remaining connected).

**Comment by mafintosh** [comment]
@joehand what about this? 

  * `dat share <link-or-folder>` (similar to git). this shares the dat and keeps downloading/uploading data.
  * `dat get <link-or-folder>` is just for downloading.  it'll fetch the data and exit. if you dat get <folder> it'll fetch updates
  * `dat <link-or-folder>` is just a shortcut to share.

(for now share will require the 2nd destination argument as you describe above when using `dat <link>` but we can make the optional in the future like we've discussed)

File: tests/download.js [hunk 1]
```diff
@@ -1,134 +1,127 @@
+var fs = require('fs')
 var os = require('os')
 var path = require('path')
 var test = require('tape')
-var rimraf = require('rimraf')
 var mkdirp = require('mkdirp')
```
Comment (mafintosh) [inline tests/download.js]: mkdirp is missing from devdeps

File: cli.js [hunk 2]
```diff
@@ -20,17 +19,98 @@ if (args.version) {
...
+    args.level = level(args.datDb)
+    run()
+  } else {
+    if (isDownload) args.dir = '.' // resume download in cwd
+    resume()
+  }
+
+  function resume () {
+    var db = args.level = level(args.datDb)
+    db.get('dat!key', function (err, value) {
```
Comment (mafintosh) [inline cli.js]: put a `!` in front of `dat` (`!dat!key`). then it'll be a valid sublevel :)

File: tests/download.js [hunk 3]
```diff
@@ -1,134 +1,127 @@
...
-var spawn = require('./helpers/spawn.js')
+var rimraf = require('rimraf')
+var spawn = require(path.resolve(path.join(__dirname, 'helpers', 'spawn.js')))
```
Comment (mafintosh) [inline tests/download.js]: this is the same as `require('./helpers/spawn.js')`

**Comment by joehand** [comment]
> * dat share \<link-or-folder> (similar to git clone / pull / push in terms or arguments). this shares the dat and keeps downloading/uploading data.
* dat get \<link-or-folder> is just for downloading. it'll fetch the data and exit. if you dat get <folder> it'll fetch updates and exit
* dat \<link-or-folder> is just a shortcut to share.

 `dat <link/folder>` is good and like the idea. But `dat share <link/folder>` will be super counter-intuitive for downloads. Though I guess if we have `dat get <link>` then thats the easier option.

**Comment by mafintosh** [comment]
@joehand if its easy to add again could you make it store the previously used port in level? like we did before.

**Comment by mafintosh** [comment]
Tests are green. I'm happy. I think we should merge this into master and use it for a bit and see how it feels. @joehand before potentially merging this could you squash the commits?

**Comment by karissa** [comment]
Yeah I'm down for merging and in the next week getting some direct feedback and see where the pain points and bugs are

File: usage/root.txt [hunk 4]
```diff
@@ -1,13 +1,22 @@
...
+dat share <directory>
+dat <directory>
+
+  create a dat-link and share directory
+
+  --snapshot            create a snapshot of directory
+  --port, -p            set a specific inbound tcp port
+
+dat get <dat-link> <directory>
+dat <dat-link> <directory>
```
Comment (karissa) [inline usage/root.txt]: why not `dat <directory> [link]` ?

**Comment by joehand** [comment]
> why not `dat <directory> [link] `?

@karissa a few reasons we went with `dat [link] <dir>`:

1. It's now `dat <source> <destination>` for both commands, keeping option to make destination optional in the future without breaking. Source will always be required.
2. `dat 4f36c088e9687ddf53d36f785ab84c65f4d24d8c4161950519b96a57d65ae08a joe/downloads` will be much easier to share/copy command or change download directories.

**Comment by joehand** [comment]
@maxogden updated with your output suggestions. 

I'll smush commits and merge tomorrow (CC: @mafintosh, @karissa).

**Comment by mafintosh** [comment]
@joehand :tada: LGTM :tada:

**Pull request merged**